tests/test_core.py

```diff
@@ -4105,6 +4105,115 @@ def test_dylink_hyper_dupe(self):
       full = run_js('src.cpp.o.js', engine=JS_ENGINES[0], full_output=True, stderr=STDOUT)
       self.assertContained("warning: trying to dynamically load symbol '_sideg' (from 'third%s') that already exists" % dylib_suffix, full)
 
+  @needs_dlfcn
+  def test_dylink_dso_needed(self):
+    # test that linking to shared library B, which is linked to A, loads A as well.
+    # main is also linked to C, which is also linked to A. A is loaded/initialized only once.
+    #
+    #          B
+    #   main <   > A
+    #          C
+    with open('liba.cpp', 'w') as f:
+      f.write(r'''
+        #include <stdio.h>
+        #include <emscripten.h>
+
+        static const char *afunc_prev;
+
+        EMSCRIPTEN_KEEPALIVE void afunc(const char *s);
+        void afunc(const char *s) {
+          printf("a: %s (prev: %s)\n", s, afunc_prev);
+          afunc_prev = s;
+        }
+
+        struct ainit {
+          ainit() {
+            puts("a: loaded");
+          }
+        };
+
+        static ainit _;
+      ''')
+
+    with open('libb.cpp', 'w') as f:
+      f.write(r'''
+        #include <emscripten.h>
+
+        void afunc(const char *s);
+        EMSCRIPTEN_KEEPALIVE void bfunc();
+        void bfunc() {
+          afunc("b");
+        }
+      ''')
+
+    with open('libc.cpp', 'w') as f:
+      f.write(r'''
+        #include <emscripten.h>
+
+        void afunc(const char *s);
+        EMSCRIPTEN_KEEPALIVE void cfunc();
+        void cfunc() {
+          afunc("c");
+        }
+      ''')
+
+    # XXX in wasm each lib load currently takes 5MB; default TOTAL_MEMORY=16MB is thus not enough
+    self.set_setting('TOTAL_MEMORY', 32 * 1024 * 1024)
+
+    so = '.wasm' if self.is_wasm() else '.js'
+
+    def ccshared(src, linkto=[]):
+      cmdv = [PYTHON, EMCC, src, '-s', 'SIDE_MODULE=1', '-o', os.path.splitext(src)[0] + so]
+      cmdv += self.get_emcc_args() + ['-s', 'RUNTIME_LINKED_LIBS=' + str(linkto)]
+      run_process(cmdv)
+
+    ccshared('liba.cpp')
+    ccshared('libb.cpp', ['liba' + so])
+    ccshared('libc.cpp', ['liba' + so])
+
+    self.set_setting('MAIN_MODULE', 1)
+    self.set_setting('RUNTIME_LINKED_LIBS', ['libb' + so, 'libc' + so])
+    self.do_run(r'''
+      void bfunc();
+      void cfunc();
+
+      int main() {
+        bfunc();
+        cfunc();
+        return 0;
+      }
+      ''',
+                'a: loaded\na: b (prev: (null))\na: c (prev: b)\n')
+
+    self.set_setting('RUNTIME_LINKED_LIBS', [])
```

Comment:
Hmm, I was hoping that using `None` here would be the way to unset a setting but it doesn't look like that works yet.  I will create patch for that.

Reply:
I think, since `RUNTIME_LINKED_LIBS` is a list, it is more explicit to reset it back to `[]`. However, if project keepers insist on resetting with `None`, or with e.g. imagined `.reset_setting()` - that can be done.